components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/DefaultURLResolverService.java

```diff
@@ -0,0 +1,173 @@
+        }
+        return hostName;
+    }
+
+    private static void appendContextToUri(String endpoint, boolean addProxyContextPath, boolean addWebContextRoot,
+                                           StringBuilder serverUrl, boolean addTenantParamLegacyMode) {
+
+        String tenantDomain = IdentityTenantUtil.getTenantDomainFromContext();
+        // If ProxyContextPath is defined then append it.
+        if (addProxyContextPath) {
+            String proxyContextPath = ServerConfiguration.getInstance().getFirstProperty(IdentityCoreConstants
+                    .PROXY_CONTEXT_PATH);
+            if (StringUtils.isNotBlank(proxyContextPath)) {
+                if (proxyContextPath.trim().charAt(0) != '/') {
+                    serverUrl.append("/").append(proxyContextPath.trim());
+                } else {
+                    serverUrl.append(proxyContextPath.trim());
+                }
+            }
+        }
+
+        // If webContextRoot is defined then append it
+        if (addWebContextRoot) {
+            String webContextRoot = ServerConfiguration.getInstance().getFirstProperty(IdentityCoreConstants
+                    .WEB_CONTEXT_ROOT);
+            if (StringUtils.isNotBlank(webContextRoot)) {
+                if (webContextRoot.trim().charAt(0) != '/') {
+                    serverUrl.append("/").append(webContextRoot.trim());
+                } else {
+                    serverUrl.append(webContextRoot.trim());
+                }
+            }
+        }
+
+        if (isTenantURLSupportEnabled()) {
+            if (StringUtils.isBlank(tenantDomain)) {
+                tenantDomain = PrivilegedCarbonContext.getThreadLocalCarbonContext().getTenantDomain();
+            }
+            if (tenantDomain != null) {
+                if (serverUrl.toString().endsWith("/")) {
+                    serverUrl.append("t/").append(tenantDomain);
+                } else {
+                    serverUrl.append("/t/").append(tenantDomain);
+                }
+            }
+        }
+
+        if (StringUtils.isNotBlank(endpoint)) {
+            if (!serverUrl.toString().endsWith("/") && endpoint.trim().charAt(0) != '/') {
+                serverUrl.append("/").append(endpoint.trim());
+            } else if (serverUrl.toString().endsWith("/") && endpoint.trim().charAt(0) == '/') {
+                serverUrl.append(endpoint.trim().substring(1));
+            } else {
+                serverUrl.append(endpoint.trim());
+            }
+        }
+
+        if (!isTenantURLSupportEnabled() && addTenantParamLegacyMode && !StringUtils.isBlank(tenantDomain)) {
+            if (!MultitenantConstants.SUPER_TENANT_DOMAIN_NAME.equalsIgnoreCase(tenantDomain)) {
+                serverUrl.append("?").append(MultitenantConstants.TENANT_DOMAIN).append("=").append(tenantDomain);
+            }
+        }
+
+        if (serverUrl.toString().endsWith("/")) {
+            serverUrl.setLength(serverUrl.length() - 1);
+        }
+    }
+}
```

Comment:
Better if this method can be broken down to chunks for readability and re-structure